Make query highlighting attributes multi-selectable

oliverguenther opened the pull request:
- Add highlighting_attributes to form, use highlightable columns to decide which of those have a color (allows easy addition of e.g., custom field columns)
- Add these attributes as embedded values to the query form
- Extract multi select component for re-use
- Save and validation
- Correct parsing of selected attributes in query API
- Use IDs instead of HREFs

**TODOS**

- [x] Adopt settings to allow the same configuration (multiselect etc)
- [x] Make settings EE aware.
- [x] Make MultiToggleSelect interface clearer for the consumer to except only an array or `undefined` and not individual objects.
- [x] Extend spec for QueryRepresenter
- [x] Spec for Query
- [x] Spec for WorkPackageColumn
- [x] Spec for ParseQueryParamsService
- [x] Spec for UpdateQueryFromParamsService

**Out of scope**
- Re-use the multi-select component for the edit fields (separate PR)

https://community.openproject.com/wp/28516

oliverguenther committed:
Add multi-select to highlighting

oliverguenther committed:
Add highlightable column

oliverguenther committed:
try to save highlighted attributes

wielinde commented on db/migrate/20180903110212_add_highlighting_to_query.rb:
```diff
@@ -1,5 +1,6 @@
 class AddHighlightingToQuery < ActiveRecord::Migration[5.1]
   def change
     add_column :queries, :highlighting_mode, :text
+    add_column :queries, :highlighted_attributes, :text
```
@oliverguenther I already had run that migration before. So that new column wasn't added when I I ran `rake db:migrate` on my machine. I will extract it to a separate migration.

wielinde committed:
Move adding "highlighted_attributes" to a separate migration as some people might have run the first migration already.

wielinde commented on frontend/src/app/components/wp-query/url-params-helper.ts:
```diff
@@ -86,6 +86,10 @@ export class UrlParamsHelperService {
     }
 
+    if (query.highlightingMode === 'inline' && query.highlightedAttributes) {
+      paramsData.hla = query.highlightedAttributes.map(el => el.href);
```
@oliverguenther Do we really want to use the `href` as attribute identifiers? Can't we keep it shorter?

oliverguenther replied:
Yes, they are the HAL identifiers of the resource. However We can map it here to its ID and unwrap it to `/api/v3/columns/ID` again in the URLParamsHelper when sending to the query endpoint, since this point here is used for the URL query string and should be concise

oliverguenther committed:
Implement frontend part of partially selectable attributes

oliverguenther committed:
Fix correctly setting undefined while backend yields wrong empty array

oliverguenther committed:
Correctly map to single value when only one selected

wielinde committed:
remove empty line [ci skip]

wielinde committed:
Fix saving of queries with highlighted attributes.

- Use column IDs instead of their href.
- Fix validation. validates_inclusion_of was not doing the job as does not check for intersection
- Extend parameter parsing for hightlighted_attributes
- Extend contract

wielinde committed:
fix specs

wielinde committed:
Add feature spec for selecting attributes for inline highlighting.

wielinde committed:
Make highlighted_attributes be part of `valid_subset!` of query so erroneous saved queries will not fail, but saving them does.

Further:
- Ensure `to_sym` on attrbute reader
- Improve validation error message by telling which attribute is making problems.

wielinde committed:
Very basic settings for default highlighting

wielinde committed:
Cleanup basic UI for setting highlighting mode and attributes

wielinde committed:
Simplify UX of work package highlighting settings

wielinde committed:
Make WP highlighting „EE only“ in settings form

wielinde committed:
Improve Codeclimate

wielinde committed:
Cleanup return values of highlighted attributes tab and improve code readability

wielinde committed:
Make empty array of highlighted attributes (hla) get reflected in URL params and dissapear

wielinde committed:
Simpler to use interface of MultiToggleSelectComponent

- Always emit an Array. No selection is an empty Array.
- Expect an Array or undefined as initialSelection.
- Allow selecting a certain value in single select when switching from multi not having any value selected.

ulferts commented on app/assets/javascripts/settings.js.erb:
```diff
@@ -100,5 +100,30 @@ See docs/COPYRIGHT.rdoc for more details.
+        $(".settings--highlighted-attributes").show();
+      } else {
+        $(".settings--highlighted-attributes").hide();
+      }
```
```
$(".settings--highlighted-attributes").toggle(highlightingMode === "inline")
```
[is equivalent](http://api.jquery.com/toggle/)

ulferts commented on app/views/settings/_work_packages.html.erb:
```diff
@@ -34,12 +34,36 @@ See docs/COPYRIGHT.rdoc for more details.
+            <p><strong><%= t('js.work_packages.table_configuration.upsale.ee_only') %></strong></p>
+            <p><%= t('js.work_packages.table_configuration.upsale.attribute_highlighting') %></p>
+            <p>
+              <a href="https://www.openproject.org/enterprise-edition/?op_edtion=community-edition&op_referrer=settings-wp-attribute-highlighting#attribute-highlighting"
```
As we are probably referring to the enterprise-edition website a couple of times (or will be) it makes sense to extract this into a helper.

wielinde committed:
Cleanup code of HighlightingTabComponent

wielinde committed:
Simplify code for toggling visibility of highlighted attributes setting.

wielinde committed:
Extract EE hint for system settings

wielinde committed:
better CodeClimate for long method encodeQueryJsonParams

wielinde committed:
Add more highlighting specs for api v3

ulferts commented on frontend/src/app/components/wp-query/url-params-helper.ts:
```diff
@@ -63,31 +63,54 @@ export class UrlParamsHelperService {
+    paramsData = this.encodeSortBy(paramsData, query);
+    paramsData = this.encodeFilters(paramsData, query);
+    paramsData.pa = additional.page;
+    paramsData.pp = additional.perPage;
```
Thanks for cleaning up.

ulferts commented on frontend/src/app/components/wp-table/configuration-modal/tabs/highlighting-tab.component.html:
```diff
@@ -12,7 +12,7 @@
   <form>
     <p [textContent]="text.highlighting_mode.description"></p>
-    <div class="form--field -full-width">
+    <div class="form--field -full-width inline-attribute-select">
```
the `inline-attribute-select` class is added only for testing purposes. Is no other selector possible?

ulferts commented on frontend/src/app/components/wp-table/configuration-modal/tabs/highlighting-tab.component.html:
```diff
@@ -21,11 +21,17 @@
       </div>
     </div>
-    <div class="form--field -full-width">
+    <div class="form--field -full-width entire-row-switch">
```
same as above

ulferts commented on spec/models/queries/work_packages/columns/work_package_column.rb:
```diff
@@ -0,0 +1,39 @@
+#-- copyright
```
As the file has no `_spec.rb` ending, it will not be executed by running `rspec [some folder]`

ulferts commented on lib/api/v3/queries/query_representer.rb:
```diff
@@ -217,6 +217,30 @@ class QueryRepresenter < ::API::Decorators::Single
+                      ::API::Utilities::PropertyNameConverter.to_ar_name(name, context: WorkPackage.new) if name
+                    end
+
+                    represented.highlighted_attributes = columns.map(&:to_sym).compact if fragment
```
The setter is not tested by the current tests:
https://github.com/opf/openproject/blob/dev/spec/lib/api/v3/queries/query_representer_parsing_spec.rb

wielinde replied:
Done

ulferts reviewed (changes requested):
I have added comments within the code where possible.

The functionality itself is looking fine with the exception of the default setting administration which was quite hard to figure out.

We already talked about the necessity of documenting the changes in the api docs.

wielinde committed:
Remove CSS classes that were only used in feature spec

wielinde committed:
Allow indenting settings so that they look like as if they belong to the preceding sibling setting

wielinde committed:
Allow for filtering `Setting` values when they are read. This is needed for overwriting values when EE token is not available.

wielinde committed:
Add API documentation for `highlightedAttributes`

[ci skip]

wielinde committed:
Fix spec

wielinde committed:
Tiny improvements for Query API documentation.

wielinde committed:
Test query validations on `highlighted_attributes`

wielinde committed:
Add spec for QueryRepresenterParsing of highlighted_attributes

wielinde commented:
@ulferts Could you please have a look at the latest commits since your last review?

ulferts commented:
:+1:
